### ISSUE TITLE
"scenario" mode?

### ISSUE THREAD

**Issue by smarteam** [magnumripper/JohnTheRipper#2349]
This is a feature request

Imagine a situation with a series of checks needed to be done on a single file with a big set of hashes or a series of files with similar checks. For example ranged by difficulty/result level to check compliance with password policy, check against previous findings (with modification rules) and also against predefined wordlists (like known public leaks). Like one big wordlist with different sets of hashes (with varying algorithms) or one big hashlist with lots of options to try (like different masks, rules etc.).

I guess it'll be handy to eliminate those re-loads and re-deduplication checks that will be made if launching in an ordinary way step-by-step with separate john process each time. By doing the resource allocation (loading/placing in memory) and after that getting the "scenario" with a list of commands/options (from a file for example) can save the time.

Something like the thing that is made during (1) (2) (3) steps of "no options" john run but more versatile and human-controlled.

john -w=verybigdict.txt -scenario=test1.list
with test1.list containing:
-format=hash1 system1.pass
-format=hash2 system2.pass
-format=hash3 system3.pass
...

john checkme.pass -scenario=test2.list
with test2.list contaning:
-mask=XXXXXXX
-mask=YYYYYYYYY
-w=previouslyfound.txt -rules=popular_modifications
-w=verybigleak.txt
...

**Comment by solardiz** [comment]
Yes, it's been planned as an enhancement/replacement of the default hard-coded "batch mode" (those 3 steps), but I never got around to it.

**Comment by magnumripper** [comment]
> john checkme.pass -scenario=test2.list
with test2.list contaning:
-mask=XXXXXXX
-mask=YYYYYYYYY
-w=previouslyfound.txt -rules=popular_modifications
-w=verybigleak.txt

The above can be handled with some extensions to existing batch mode. I like this idea and I have looked into it before. Should be fairly easy to implement.

> john -w=verybigdict.txt -scenario=test1.list
with test1.list containing:
-format=hash1 system1.pass
-format=hash2 system2.pass
-format=hash3 system3.pass

The above would be tricky. We'd need to re-load hashes inbetween, and we wouldn't gain anything compared to normal scripting or a `john this && john that` one-liner. I'm not sure we should support it.

**Comment by smarteam** [comment]
>>john -w=verybigdict.txt -scenario=test1.list
>>with test1.list containing:
>>-format=hash1 system1.pass
>>-format=hash2 system2.pass
>>-format=hash3 system3.pass
>
> The above would be tricky. We'd need to re-load hashes inbetween, and we wouldn't gain anything compared to normal scripting or a john this && john that one-liner. I'm not sure we should support it.

It was just an example, maybe more real-worldish scenario will be a dump from database of some web service that changed hashing algorithm but still store old values. So it is a single file with similar (or even same) passwords in different formats and possibility to use a --loopback can be handy I guess.

But totally the point was about having some static common part able to work with configurable input batches, glad to see the support for idea.

**Comment by magnumripper** [comment]
We could actually implement #2310 as a special case of this batch mode (just a different input file format).

**Comment by AlekseyCherepanov** [comment]
When john loads a lot of hashes (millions), it is beneficial to be able to run several attacks without unloading hashes. That's quite realistic scenario.

OTOH I thought about it not as a mode, but more as a server: john loads hashes and starts a server, so a client may connect and run several attacks, maybe reconnect later and run more. Reconnection is important because it would be easier to develop client this way.

Also server would allow non-linear execution of scenarios based on feedback. But then it would be client's responsibility to provide save-restore, while john would save only progress during 1 attack.
